### PR TITLE
chore(repo): add Vitest into codebase

### LINKED ISSUE #44: [ADR] Test Driven Development Philosophy
**Problem Statement**
If you've ever worked with a large code base you will understand the frustration that comes with not understanding your peer's code. Code can get messy and inconsistent, in this jumble its not uncommon to misunderstand the absolute base requirements of a feature or API endpoint. This is where Test Driven Development (TDD) comes in. 

Test are written first, then we write the minimum code to needed to pass the test, after tests pass, we begin to refactor the code. 
image

** Why is this important **
- Documentation: Test outline the basic requirements of any feature or endpoint. To put it simply, our tests are our documentation. Any developer can read the tests we wrote and understand what exactly a feature should be doing.
- Promotes collaboration: Since our tests are our documentation, any developer can modify existing code since they know exactly what tests need to pass for their improvements to be successful.
- Development speed: Initial development speed may be slower, but with well written tests we will be able to ship out code and new features at higher and higher speeds. 
- Higher quality code: Writing code that passes tests encourages developers to write more modular and incremental code that is higher in quality overall. 

**Goals**
Goal 1: Write tests before any feature
Goal 2: Testing on frontend components and backend endpoints
Goal 3: Anyone can understand what our codebase does
Goal 4: Develop a robust code base with minimal bugs and a "fail-fast" approach
Goal 5: Integrate unit testing with Vitest

**Out of Scope**
- Excessive mocking of data to test smaller functions, we should integration tests here
- Performance testing
- Testing non requirements for our unit tests

## Solution Summary

**History/Context**
Historically, I've done development that stalls out near the end of development because of constantly needing to refactor old code. With this approach we can avoid this and provide a stable ground for all development.

**Solution**
- Vitest will be used primarily for unit testing

**Other options considered**
Other philosophies considered: Behaviour driven development. BDD doesn't allow us to be as modular and incremental as TDD does, so it was opted to use TDD.
Other tools considered: Jest. Vitest does everything that Jest does but better, Jest is not well suited for modern javascript applications. Vitest also integrates better with Next.js and is well maintained by the Vite team.

### PR DESCRIPTION
**Pull request by HasithDeAlwis**
## Describe The Outcome
- Added scaffolding of Vitest into codebase, this lays down foundation for creating new unit-tests with Vitest throughout our code base
- Added a workflow file to run all unit-tests on each pull-request

## Dependency Changes
- vitest v2.0.5 installed

## Context
Essential for following Test Driven Development philosophies. 

closes #44

### CODE REVIEW

File: eslint.config.mjs [hunk 1]
```diff
@@ -8,5 +8,6 @@ export default antfu(
   },
   {
     ...playwright.configs['flat/recommended'],
+    ignores: ['tests/unit-tests/**/*.ts'],
```
Comment (HasithDeAlwis) [inline eslint.config.mjs:11]: I'm aware this is causing a build failure. Let me know how to go about this if you have any advice. Currently looking into a solution.

**Comment by MFarabi619** [comment]
The branch for [this](https://github.com/cuhacking/hackathon/pull/57) is currently `hasith/44-test-driven...` which is good, I can tell who made it and what issue it was for. To make it better you can do `hasith/test/44-test-driven...` or `hasith/chore/44-test-driven...` since it was just scaffolding Vitest rather than actually writing tests. The branching guidelines are mentioned [here](https://deploy-preview-54--cuhacking-portal-test-deployment.netlify.app/docs/contribution-guidelines/branching-guidelines) in the docs. Feel free to add to or change it if you think it could be clearer.

Sometimes which type to apply can be ambiguous, so use your best judgement, can always be changed later. If it's getting ambiguous, it's also a good sign to split the issue into separate smaller issues to separate concerns. The issue title should specify both the branch name and the PR title. 

The issue title should also be in imperative tense, so `chore: Scaffold Vitest ....`. And then your branch would be `hasith/chore/44-scaffold-vitest...` and the PR title would be the same as the issue title, `chore: Scaffold Vitest ...`. See how much easier that makes things?

Just like there's a commit history....
image

There's also branching history....
image

Adhering to conventional spec also allows you to switch between branches easier when multi-tasking, so you can take better advantage of the distributed nature of Git. You know who's branch it is, what they're working on, and what type of work it is.

People looking at the list of issues can also better make out what to do. So for example this:
image

Could be `[ADR] Add Issue Creation Feature to Github-Discord Webhook to Improve Communication`

Since it's not a huge task, it could just be a feature issue instead. ADRs are for like big project decisions like what tech to use, what philosophy to adapt, etc.

No need to change the branch for this time. Just a note for the future, this comment can also be sent as a link to others in case this happens again.

**Comment by HasithDeAlwis** [comment]
Thanks for the advice on this, just completely missed my brain to add the spec, nothing wrong with the docs on that one. I renamed the other issue to be clearer 👍

**Comment by MFarabi619** [comment]
All good all good we're learning

**Comment by MFarabi619** [comment]
image

This property doesn't exist for [eslint-plugin-playwright](https://github.com/playwright-community/eslint-plugin-playwright).

Were the playwright ESLint rules causing issues with Vitest tests?

File: eslint.config.mjs (hunk 1 above)
Comment (MFarabi619) [reply eslint.config.mjs:11]: image

This property doesn't exist for [eslint-plugin-playwright](https://github.com/playwright-community/eslint-plugin-playwright).

Were the playwright ESLint rules causing issues with Vitest tests?

Comment (HasithDeAlwis) [reply eslint.config.mjs:11]: Screenshot 2024-08-12 at 6 47 59 PM
ESLint misidentifies the Vitest code as playwright

**Comment by MFarabi619** [comment]
image

In your `tests` directory, you added `unit-tests`. This is unnecessary because it's already inside the `tests` directory. You can just call it `unit`. Later you'll have `integration`.

So we eventually end up the following:
image

You can see the `docs` goes against our system. It signifies that these are tests for the docs site but we don't know what type. Only after you open it and see `pom.ts`

image

You can tell it was e2e tests. This will not scale. Hence why e2e tests are split into a separate app in our [upcoming monorepo architecture](https://github.com/cuhacking/hackathon/issues/40) as they do need maintenance and can grow into a significant amount of code.

File: eslint.config.mjs (hunk 1 above)
Comment (MFarabi619) [reply eslint.config.mjs:11]: Yep working on fixing it currently

**Comment by HasithDeAlwis** [comment]
Would unit-tests and integrations also be their own app? 
Really cool that we can split up our apps to have their own E2E testing.

**Comment by MFarabi619** [comment]
> Would unit-tests and integrations also be their own app? Really cool that we can split up our apps to have their own E2E testing.

Depends on scale. If your unit and integration test suites should be co-located with your app. They generally don't take too long to run, but have the highest maintenance cost. If your unit and integration tests are big enough to warrant being separated, your app should be split up/re-architected instead.

E2E tests take a while and have the lowest maintenance cost. They should also be completely black-boxed from your main app, knowing absolutely nothing about it. Hence why it gets its own directory.

File: eslint.config.mjs (hunk 1 above)
Comment (MFarabi619) [reply eslint.config.mjs]: Couldn't fix it, commented it out altogether for now. The E2E tests will be in a separate directory in the monorepo so we won't have conflicting configs anyways. Monorepo management tools handle a lot of these config headaches for you.

**Review by MFarabi619 — APPROVED** [review]

File: .github/workflows/TEST_unit.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      - name: 🍞 Install Bun
+        uses: oven-sh/setup-bun@v2
+        with:
+          bun-version: latest
+
+      - name: 📂 Install Dependencies
+        run: bun install
+
+      - name: 💀 Run Unit Tests
+        run: bun run test:vitest
```
Comment (HasithDeAlwis) [inline .github/workflows/TEST_unit.yml:32]: make sure we rename this accordingly as well

Comment (MFarabi619) [reply .github/workflows/TEST_unit.yml:32]: Doing some important work and can't stash, feel free to change this and force push

Comment (MFarabi619) [reply .github/workflows/TEST_unit.yml:32]: Also, you want you keep the commands generic, makes mass migrations much easier. So the script for `test:unit` could use Jest in the future if necessary.

**Comment by JowiAoun** [comment]
Looks good, as well as Farabi's comments here. Awaiting permissions on the deployment, I would be happy to take a chance at fixing the pipelines.

**Review by JowiAoun — APPROVED** [review]

**Pull request merged**